### PR TITLE
Added structs and basic methods for handling Slack dialogs

### PR DESCRIPTION
**Pull request by robiball**
I've not had a chance to add proper testing, but I can try and add some in the next week or so if you want it.

### CODE REVIEW

**Review by upgreydd — APPROVED** [review]

**Review by ruudk — APPROVED** [review]
🙌

**Comment by ruudk** [comment]
Works great.... How do you easily parse the interactive component callback into a slack.DialogCallback?

**Comment by robiball** [comment]
It's pretty easy, if you convert the slack request body to JSON, you can just do this:

`var dialog slack.DialogCallback`
`err := json.Unmarshal([]byte(jsonStr), &dialog)`

I was doing something similar to this at the time: https://medium.com/mercari-engineering/writing-an-interactive-message-bot-for-slack-in-golang-6337d04f36b9
Which gives a nice example on how to get the `jsonStr` from the slack request :)

**Comment by ruudk** [comment]
@robiball I just came to the same conclusion. I was thinking too complicated. Thanks for the fast reply!

**Comment by ruudk** [comment]
@robiball Can you rebase your master branch against nlopes/slack master? It doesn't contain slackevents.

Edit: nvm handled it in my own fork.

File: dialog.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+type DialogElementOption struct {
+	Label  string   `json:"label"`	//Required.
+	Value  string   `json:"value"`	//Required.
+}
+
+// DialogCallback is sent from Slack when a user submits a form from within a dialog
+type DialogCallback struct {
+	Type         string             `json:"dialog_submission"`
```
Comment (kbvincent) [inline dialog.go]: would this not be `json:"type"`?

Comment (robiball) [reply dialog.go]: Whoops... Nice spot. I'll update it now.

Comment (robiball) [reply dialog.go]: Done.

**Comment by kbvincent** [comment]
Good work @robiball!
Maintainers: Would love to see this merged so we dont have to use a fork.

File: misc.go [hunk 2]
```diff
@@ -154,6 +147,27 @@ func postForm(ctx context.Context, client HTTPRequester, endpoint string, values
 	return parseResponseBody(resp.Body, intf, debug)
 }
 
+func postJson(ctx context.Context, client HTTPRequester, endpoint, token string, json []byte, intf interface{}, debug bool) error {
+	reqBody := bytes.NewBuffer(json)
+	req, err := http.NewRequest("POST", SLACK_API+endpoint, reqBody)
```
Comment (james-lawrence) [inline misc.go]: lets keep this method consistent with postForm below and only accept the fully formed URL.

**Review by james-lawrence — CHANGES_REQUESTED** [review]

**Comment by james-lawrence** [comment]
will accept with a minor change. tests would be good as well.

File: dialog.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	resp := DialogTrigger{
+		TriggerId:  triggerId,
+		Dialog:     dialog,
+	}
+	jsonResp, _ := json.Marshal(resp)
+	response := &SlackResponse{}
+	if err := postJson(ctx, api.httpclient, "dialog.open", api.token, jsonResp, response, api.debug); err != nil {
+		return err
+	}
+	if !response.Ok {
```
Comment (james-lawrence) [inline dialog.go]: SlackResponse has a method for this.

**Comment by robiball** [comment]
Thanks @james-lawrence. I've made those updates, I'll get some tests added shortly and update the PR.

**Comment by nitsanfiverr** [comment]
@robiball Anything new with the PR ?

**Comment by james-lawrence** [comment]
@nitsanfiverr  if you want to finish up the changes and open another PR I'll merge it in.

**Comment by robiball** [comment]
Apologies for the delays, I've been incredibly busy over the last couple of weeks and haven't had a chance to add the tests.

All done now, and updated the PR. Hope it's okay!

File: dialog.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+func (api *Client) OpenDialogContext(ctx context.Context, triggerId string, dialog Dialog) (err error) {
+	if triggerId == "" {
+		return errors.New("received empty parameters")
+	}
+
+	resp := DialogTrigger{
+		TriggerId:  triggerId,
+		Dialog:     dialog,
+	}
+	jsonResp, _ := json.Marshal(resp)
```
Comment (james-lawrence) [inline dialog.go]: i know its unlikely but lets not ignore marshalling errors.

**Review by james-lawrence — CHANGES_REQUESTED** [review]
soo close, sorry if i missed this in the last PR. but don't ignore errors. makes them a bear to track down.

File: dialog.go (hunk 4 above)
Comment (robiball) [reply dialog.go]: No problem, updating it now!

Comment (robiball) [reply dialog.go]: Done!

**Comment by robiball** [comment]
@james-lawrence updated with your comment :)

**Review by james-lawrence — APPROVED** [review]

**Pull request merged**